Clear up the text for converge_by

EasyAsABC123 opened the pull request:
Importantly removed it would `touch the file` which is what doesn't happen during a dry run.
To make it clear that `converge_by` doesn't deal with idempotency and is instead useful for `why-run`.

Peer-documented @iennae
Review Requested to @lamont-granquist

chef-delivery commented:
This PR has passed 'Verify' and is ready for review and approval!
  Use: '@delivery approve' when code review is complete.

jamescott commented:
@delivery approve

chef-delivery commented:
Change: ad88f3fe-d8a8-4ab8-9194-255adae4e207 approved by: @jamescott

chef-delivery commented:
This PR has passed 'Acceptance' and is ready to be delivered!
  Use: '@delivery deliver' when validated in acceptance.

EasyAsABC123 committed:
Clear up the text for converge_by 

Importantly removed it would `touch the file` which is what doesn't happen during a dry run.
To make it clear that `converge_by` doesn't deal with idempotency and is instead useful for `why-run`.

Peer-documented @iennae
Review Requested to @lamont-granquist

lamont-granquist commented on chef_master/source/custom_resources_notes.rst:
```diff
@@ -181,9 +181,9 @@ If you do need to write code which mutates the system through pure-Ruby then you
    end
 
-The ``converge_by`` block gets why-run correct and will just touch "/tmp/foo" instead of actually doing it. The ``converge_by`` block is also responsible for setting ``update_by_last_action``.
+The ``converge_by`` block gets why-run correct and will print to ``Chef::Log.info`` "touch /tmp/foo" and not run ``::FileUtils.touch "/tmp/foo"``. The ``converge_by`` block is also responsible for setting ``update_by_last_action``.
```
This is awkward reading.

The prior text was incorrect and mangled somehow.

Maybe:

```
When the ``converge_by`` block is run in why-run mode, it will only log ``touch "/tmp/foo"`` and will not run the code inside the block.
```

I really dislike throwing the `Chef::Log.info` and `FileUtils.touch` details at the reader.  It is also likely better to be clear that the code inside the block is not executed -- there is no magic other than the block simply is not run.  Also "gets why-run correct" is awkward (i think I wrote that one initially, it sounds like me) and should really be changed.

EasyAsABC123 committed:
Update custom_resources_notes.rst

EasyAsABC123 commented:
@lamont-granquist agreed, let me know if the new text makes more sense :) i only added the `converge_if_changed` part as i thought it was close to the same, i now know i am wrong.

lamont-granquist commented on chef_master/source/custom_resources_notes.rst:
```diff
@@ -181,9 +181,9 @@ If you do need to write code which mutates the system through pure-Ruby then you
    end
 
-The ``converge_by`` block gets why-run correct and will just touch "/tmp/foo" instead of actually doing it. The ``converge_by`` block is also responsible for setting ``update_by_last_action``.
+When the ``converge_by`` block is run in why-run mode, it will only log ``touch "/tmp/foo"`` and will not run the code inside the block. 
```
oops I didn't mean for the sentence "The ``converge_by`` block is also responsible for setting ``update_by_last_action``." to get dropped as well, that should probably get re-added...

EasyAsABC123 replied:
@lamont-granquist commented that it was moved down to the next paragraph, as i felt it made more sense in context with idempotency

lamont-granquist replied:
yes, actually i see that now...

EasyAsABC123 replied:
@lamont-granquist I changed it to API but i feel like it should be `block`

EasyAsABC123 commented on chef_master/source/custom_resources_notes.rst:
```diff
@@ -181,9 +181,9 @@ If you do need to write code which mutates the system through pure-Ruby then you
+When the ``converge_by`` block is run in why-run mode, it will only log ``touch "/tmp/foo"`` and will not run the code inside the block. 
 
-The ``converge_by`` block does not deal with idempotency and will set the ``update_by_last_action`` to ``true`` everytime the block executes.  By wrapping this block with an idempotency check like ``converge_if_changed`` or ``unless File.exist?("/tmp/foo")`` this will stop ``update_by_last_action`` from being set.
+The ``converge_by`` block is also responsible for setting ``update_by_last_action``, however it does not deal with idempotency and will set the ``update_by_last_action`` to ``true`` everytime the block runs.  By wrapping this block with an idempotency check like ``converge_if_changed`` or ``unless File.exist?("/tmp/foo")`` this will stop ``update_by_last_action`` from being set.
```
@lamont-granquist not dropped moved down here

lamont-granquist commented on chef_master/source/custom_resources_notes.rst:
```diff
@@ -181,9 +181,9 @@ If you do need to write code which mutates the system through pure-Ruby then you
+When the ``converge_by`` block is run in why-run mode, it will only log ``touch "/tmp/foo"`` and will not run the code inside the block. 
 
-In order to use ``converge_by`` correctly you must ensure that you wrap the ``converge_by`` with an idempotency check otherwise your resource will be updated every time it is used and will always fire notifications on every run.
+The ``converge_by`` block is also responsible for setting ``update_by_last_action``, however it does not deal with idempotency and will set the ``update_by_last_action`` to ``true`` everytime the block runs.  By wrapping this block with an idempotency check like ``converge_if_changed`` or ``unless File.exist?("/tmp/foo")`` this will stop ``update_by_last_action`` from being set.
```
This is a run-on sentence, and you've also dropped the fact that it fires notifications on every run when you forget it.   Also `converge_if_changed` wraps a `converge_by` block as well, so those are mutually exclusive, you shouldn't use both.   This should probably read more like:


    The ``converge_by`` block does not do any checking for idempotency and always sets ``updated_by_last_action``.  A
    ``converge_by`` block that is not wrapped in an idempotency check will always cause the resource to be updated, and
    will always cause notifications to fire.  A properly written resource should wrap all ``converge_by`` checks with an
    idempotency check, or the ``converge_if_changed`` API should be used instead.   The ``converge_if_changed`` API
    wraps a ``converge_by`` block with an idempotency check for you.

EasyAsABC123 replied:
@lamont-granquist i didn't know `converge_if_changed` wrapped `converge_by` that should be in the other docs.  Awesome!

lamont-granquist replied:
some awkward code reading:

https://github.com/chef/chef/blob/master/lib/chef/provider.rb#L269,L285

all the code paths there wind up with the block getting passed into `converge_by`

EasyAsABC123 replied:
@lamont-granquist thanks for that! i was using these all wrong, greatly appreciated.

EasyAsABC123 committed:
Update custom_resources_notes.rst

EasyAsABC123 committed:
Update custom_resources_notes.rst

lamont-granquist commented on chef_master/source/custom_resources_notes.rst:
```diff
@@ -181,9 +181,13 @@ If you do need to write code which mutates the system through pure-Ruby then you
+``converge_by`` block that is not wrapped in an idempotency check will always cause the resource to be updated, and
+will always cause notifications to fire.  A properly written resource should wrap all ``converge_by`` checks with an
+idempotency check, or the [``converge_if_changed``](https://github.com/chef/chef-web-docs/blob/master/chef_master/source/custom_resources.rst#converge_if_changed) block should be used instead.   As the ``converge_if_changed`` API
+wraps a ``converge_by`` block with an idempotency check for you.
```
There's a dangling sentence fragment here, and I think there needs to be a connection between the negative effects (notifications always firing), to the solution (wrap in idempotency), and converge_if_changed then probably needs to stand alone as its own sentence.  Something like:

```
The ``converge_by`` block does not do any checking for idempotency and always sets ``updated_by_last_action``. 
A ``converge_by`` block that is not wrapped in an idempotency check will always cause the resource to be updated,
and will always cause notifications to fire.  To prevent this, a properly written resource should wrap all
``converge_by`` checks with an  idempotency check.  The
[``converge_if_changed``](https://github.com/chef/chef-web-docs/blob/master/chef_master/source/custom_resources.rst#converge_if_changed)
block may be used instead  which will wrap a ``converge_by`` block with an idempotency check for you.
```

lamont-granquist replied:
^ that edit has some redundancy but is getting closer to it

lamont-granquist commented:
this is basically good to go technically :+1:

Commit:
Very light copy editing

kagarmoe commented:
![tenor-31273509](https://cloud.githubusercontent.com/assets/4400151/26808687/76d4db74-4a13-11e7-9a88-e2ad680160dd.gif)
